### ISSUE TITLE
Switch from KiB/MiB to KB/MB in order to match CLI better?

### ISSUE THREAD

**Issue by andreasn** [cockpit-project/cockpit#704]
Opening this can of worms...
If one is working with both Cockpit and the CLI it might be odd that one displays the storage units as MiB and the other as MB.

**Comment by dperpeet** [comment]
Reopening based on comments by @cmurf in #1562

**Comment by jscotka** [comment]
Hi,
My suggestion is to add info to help for event  of mouse 
show valuses in bytes and help what kB or KB or KiB mean 
http://en.wikipedia.org/wiki/Kilobyte
http://en.wikipedia.org/wiki/Kibibyte
     Regards
     Honza

**Comment by cmurf** [comment]
I think the reference to my post in #1562 is sufficient rather than copy/pasting it here. If it warrants further discussion then we can do it here. I'm not really sure what about this is all that controversial.

If I have a distance of 10 miles I do not write that out as "10 km". That would be a bad idea. But effectively that's what change proponents are arguing - "well everyone else does it wrong, so Cockpit should do it wrong too so that it's right." Huh? Umm, no.

Metric correlates to the manufacturer usage. Smartmontools follows metric also, and they denote with metric units. So that's OK:
User Capacity:    26,843,545,600 bytes [26.8 GB]

The trend elsewhere is using binary values, and shifting to correctly labeling them with the binary unit designation (MiB, GiB, TiB) instead of metric.

Some think they can bypass this by using neither set of units and just using a letter to requires a decoder ring in documentation to understand what the quantity under discussion is. And I think that's wrong too, but marginally less wrong than using flat out the wrong unit.

**Comment by julienlim** [comment]
I've read both this and #1562 threads.  First and foremost, we definitely want consistency.  Secondly, for any storage-related values shown, the expected behavior by users is to see them in the KB, MB, GB, TB, etc. metrics.  This is common practice in storage products throughout the industry, e.g. EMC, IBM, HDS, HP, NetApp, Microsoft, VMware, etc.

If there was discrepancy in the conversion in order to get to properly calculate or convert into the the n-byte prefix (i.e. KB, MB, GB, TB), then that should be addressed.

While technically, there's nothing wrong with the metrics MiB, GiB, TiB, that's not commonly used for metrics for storage reporting.  No decoder ring needed as this is commonly expected and known by users in general.

**Comment by mvollmer** [comment]
Let's go back to the example. A possible value for a mdraid chunk size is 524288 bytes, for exampe. We let the user pick this from a dropdown. What text should be in that dropdown?

524288 B
512 KiB
524 kB
0.5 MiB
0.524 MB
512 KB
512 kB
0.5 MB

**Comment by cmurf** [comment]
tl;dr  I still can't conclude definitively whether day to day users or
sysadmins care about bytes (implies base 10 units), or fs block and file
sizes (implies base 2).

Data points:

SNIA's 2014 Dictionary says: "The SNIA uses the base 10 convention commonly
found in I/O-related and scientific literature rather than the base 2
convention (1,024, or 210) common in computer system and software
literature."

Clearly they're suggesting their base unit of interest is the byte, and
thus base 10 makes it convenient to do that. But they're also incentivized
to use base 10 because it inflates storage capacities relative to two other
contexts:

1. The device sector size is either 512 bytes or 4096 bytes, both of which
are effectively base 2 units (512 byte is describable as a simple fraction
in base 2; a complex fraction is needed to describe it in base 10). This is
the case on all platforms everywhere.

2. On Linux, the memory page size on x86 is fixed at 4096 bytes or 4KiB.
And the minimum fs block size is set to the page size. Therefore human
readable file sizes are always described using base 2 units because they
always take up a minimum space of 4KiB and in 4KiB increments.

Tools:

mdadm, LVM, and Btrfs  folks are using base 2 / binary prefixes. I'm not
sure about ZFS folks off hand. But I'm not sure how to rectify this with
storage vendors using base 10. I don't know which unit helps the sysadmin
do their job better. It could literally be miles vs kilometers, and is
personal preference and familiarity. If so that suggests a toggle may be
needed. But certainly at the very least the correct labels need to be used
regardless.

I'm not sure how to rectify that with the storage vendors using base 10. I
don't know which unit helps a sysadmin do their job better, it might
literally be a miles vs kilometers thing and if so that suggests maybe a
toggle is needed in the UI eventually.

Error:

There is a lot of misuse of SI prefixes which means it's probably not as
trustworthy as binary prefixes. People who care will really wonder "It says
KB, but do they mean KiB?"  There's essentially no one fibbing or making
the mistake of calling base 10 unit values as KiB. But there's a lot of
base 2 unit values reported as KB.

Example: mdadm-3.3.2 documentation:
-c, --chunk=
Specify chunk size of kibibytes.  The default when creating an array is
512KB.  To ensure compatibility with earlier  versions,  the default when
building an array with no persistent metadata is 64KB.

Clearly the first statement can be proven to be correct. The second and
third statement units are incorrect they are KiB not KB.

Contexts:

Point of sales / design is clearly base 10.
Creation of the storage stack is clearly base 2.
Day to day usage appears to be mixed depending on the UI you use.

**Comment by mvollmer** [comment]
> But there's a lot of base 2 unit values reported as KB.

This is what Cockpit does, unfortunately.  From `base/cockpit.js`:

```
    var byte_suffixes = {
        1024: [ null, "KB", "MB", "GB", "TB", "PB", "EB", "ZB" ],
        1000: [ null, "KB", "MB", "GB", "TB", "PB", "EB", "ZB" ]
        /* 1024: [ null, "KiB", "MiB", "GiB", "TiB", "PiB", "EiB", "ZiB" ] */
    };

    cockpit.format_bytes = function format_bytes(number, factor, separate) {
        if (factor === undefined)
            factor = 1024;
        return format_units(number, byte_suffixes, factor, separate);
    };

    var byte_sec_suffixes = {
        1024: [ "B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB" ]
    };

    cockpit.format_bytes_per_sec = function format_bytes_per_sec(number) {
        return format_units(number, byte_sec_suffixes, 1024, false) + "/s";
    };
```

**Comment by julienlim** [comment]
There are definitely a lot of differing opinions here.  To answer Marius' question on what to show for mdraid chunk size, IMHO I still believe it should show as 512 KB.

There is misuse of SI prefixes, and although International Electronic Commission established the term kibibyte for 1024 bytes, the term never caught on.  If you look around you, you'll find that very few use the terms kibibyte and KiB.  Many documents on the Internet which talks about memory and storage uses the terms kilobyte/KB, megabyte/MB, gigabyte/GB, etc. You have to go out of your way to find people who use the terms kibibyte/KiB, mebibyte/MiB, gibibyte/GiB, etc.  In other words, most of the computing industry has ignored the guidance of the IEC.

My recommendation is to follow prevailing widespread practice to align more closely with what customers' / users' expectations are as doing anything different is not only an outlier but would then make the users question other aspects about the product which can impact brand, trust, and usability.

**Comment by cmurf** [comment]
@julienlim
Every single thing you said is incorrect.

mdadm chunk size is in kibibytes, it says this in the man page; that subsequent examples use the wrong labeling only proves my point that this is a problem that causes confusion. If you look at SNIA's DDF 2.0 spec, strip_size=chunk=stripe_depth are in base 2 units. That they wrongly use SI prefixes is not a good reason for everyone else to wrongly use them too.

The IEC term never caught on? Except that it's adopted by every major standards organization on the planet, not just the IEC? NIST requires SI prefixes for decimal applications only. The IEC units are incorporated into the International System of Quantities, just like SI units are.

Plumbers have no need to distinguish between quarts and liters, the difference is minor. They're still professional. Comedians correctly make fun of people like me for being fixated on words that have meaning for the sake of accuracy, because they also have no other need to distinguish between quarts and liters. It's fine if you want to be included in the camp of plumbers, comedians, and everyone else who don't actually care about accuracy, yet are still professionals.

But it is unprofessional to give bad advice to people who do want to be accurate. You're arguing against international standards, and are injuring the efforts of people trying to eliminate the confusion and ambiguity caused by lazy compsci people who wrongly usurped an existing prefix to mean something it plainly doesn't.

I don't see how you talk about trust and usability while recommending people willfully use the wrong unit labels. Please, please, please stop doing that. I invite you to instead resort to ad hominem attacks against me personally, call me names, make a joke, whatever you want. That is far less damaging than the junk excuses you're recommending.

**Comment by julienlim** [comment]
@cmurf 

This conversations seems to have morphed into something more than intended.

What I was trying to convey and perhaps did not state it  well (and my apologies if the message got misconstrued along the way) is that customers expect to see the unit label of KB, MB, GB, etc. and not KiB, MiB, GiB, etc. and this is something widely adopted in the industry today -- whether correctly or not (and my intention is not to debate this point).

I was not talking accuracies and understand and acknowledge today there is discrepancies in how the values are computed and shown in accordance with the unit label shown.  My recommendation and hope is that the values be calculated correctly (to match that of the unit label) and show with the unit label of KB, MB, GB, etc.

**Comment by cmurf** [comment]
@julienlim 
"Widely adopted" precisely where? Be exact. What interface? What context?

Btrfs, by default, uses IEC units and properly labels them KiB, MiB, GiB, TiB

LVM, by default, uses IEC  units and properly labels them KiB, MiB, GiB, TiB, except in lvs/pvs/vgs where it uses m, g, or t (lower-case) as short hand.

ls and df, use bytes, with -h it uses IEC values but with ambiguous labeling K, M, G, T, where K is always wrong because that's the symbol for the kelvin.

Near as I can tell, base 10 values are used in marketing materials when the context is the raw storage capacity of the device in bytes. Bytes are instantly useless, though, because the smallest actual unit for storage isn't the byte, it's the block or sector size which is always a base 2 / IEC unit.

The sane reason for a UI showing base 10 / SI  units (value+label) is for device identification. People will want to see a device statistic in SI units because that's what's on the label of the device, and in a spec sheet. But no where else is that used by a storage sysadmin.

And the actual end user? They don't know enough to even care until the sizes in question become huge, and then they start to wonder, maybe a little bit.

And I will bet any storage sysadmin looking at files and filesystem stats, will want to see things in IEC units, because all of it down to the hardware is base 2 units. So if he's going to be shown base 2 units, that means IEC labeling with KiB, MiB, GiB, TiB, PiB.

Only when the sysadmin writes up a purchase order for more storage will they need this daft conversion to base 10, and mainly that's because storage vendors are jerks.

Even Apple are jerks about this in their UI. diskutil info on the command line says this about the currently mounted volume:
   Total Size:               209.7 GB (209723797504 Bytes) (exactly 409616792 512-Byte-Units)
   Volume Free Space:        106.9 GB (106912116736 Bytes) (exactly 208812728 512-Byte-Units)


df -h says this:
$ df -h
Filesystem      Size   Used  Avail Capacity  iused    ifree %iused  Mounted on
/dev/disk1     195Gi   96Gi  100Gi    49% 25105096 26097001   49%   /

And df -h is closer to correct, because inherent to using base 2 IEC units is the fact that the unused portion of each sector is counted against used; whereas in base 10 SI units there's a bit fat lie going on implying unused sector space is volume free space. Due to the fact the actual minimum unit size if 4096 bytes because that's the file system unit, the actual available capacity is probably yet again quite a bit less than the 100GiB reported by df, let alone the 107GB reported by diskutil. Which does Apple show in the OS X UI? 106.91GB — using sig figs as if it's actually an accurate assessment rather than complete b.s.

Why? Well, Apple's a storage vendor. So it makes sense they lie in their own UI.

**Comment by mvollmer** [comment]
> My recommendation and hope is that the values be calculated correctly (to match that of the unit label) and show with the unit label of KB, MB, GB, etc.

What about this: Cockpit will use the SI and IEC units correctly and it will be consistent with other correct uses of those units.

**Comment by dperpeet** [comment]
> What about this: Cockpit will use the SI and IEC units correctly and it will aim to be consistent with other correct uses of those units.

I think this is a good summary and agree. Additionally, I agree with @cmurf and think that we should use (correct) base 2 units for storage, since that is usually the more interesting value.

I also strongly agree that "others are doing it wrong" isn't a valid argument. While not everyone may be familiar with IEC units (Windows users maybe?), the units can be looked up easily and are not up for interpretation. When you see SI units for storage you are pretty much forced to go deeper in the documentation to know how that specific software chose to use the unit.

**Comment by julienlim** [comment]
I'm fine if we stick with the base 2 units for now but please bear in mind that as we converge or integrate with other non-Linux based technologies, trying to normalize and "report" on storage will become increasingly challenging.  E.g. if you have storage from EMC, NetApp, IBM, etc. and you're using NFS or some other protocol to connect to it, when we try to unify and report on various disparate subsystems, we will need to look into how to converge and compute them so they are normalized to the same unit of measure.

I will also add that if users have to look a lot of things up and compute it in their mind or elsewhere, it makes using such a product more challenging as compared with an alternative product where you don't need to.

**Comment by cmurf** [comment]
I wrote: *Due to the fact the actual minimum unit size if 4096 bytes because that's the file system unit, the actual available capacity is probably yet again quite a bit less than...*  This is wrong. df reports used as blocks in use * block byte size. So a 64KiB fs block/cluster, a 128 byte file is reported by df as 64KiB used. SI units will always overestimate free space as a result.

OS X GUI currently uses SI units (base 10 values, SI prefix). Older versions used IEC units, weird change.

Gnome Nautilus uses SI units (and yay, kB instead of KB).

Windows 2008 Server reports base 2 / IEC based values, but labels them with SI prefixes (and Wikipedia confirms my direct finding is apparently the case for all Windows).

**Comment by stefwalter** [comment]
@dperpeet I believe this is yours.

**Comment by cmurf** [comment]
The subject is backwards. Most of the CLI tools now default to IEC/KiB/MiB units. And even the SNIA dictionary of terms overwhelmingly is predicated on either 512 bytes or 4096 bytes as base (minimum) units. It's an open question whether Cockpit's default should be in parity with CLI tools, or GUI tools though.

**Comment by vpodzime** [comment]
I think we need consistency here. [Blivet] (https://github.com/rhinstaller/blivet) and thus also [Anaconda] (https://github.com/rhinstaller/anaconda) understand both decimal (KB, MB,...) and binary (KiB, MiB,...) units and the one-char prefixes are interpreted as binary units (i.e. ``10 K`` is ``10 KiB``). The string representation of size (when shown to users, logged,...) always uses binary units because we think those are clearer and unambiguous. Moreover, they more often produce nice numbers, because as mentioned above, 512 bytes or 4096 bytes are bases these days and ``512 B`` **is not really** ``0.5 KB`` while it **is** ``0.5 KiB``. ``512 B`` is the "logical" sector size used by the kernel for all storage sizes (and all real-world sector sizes are multiples of ``512 B`` AFAICT).

Interpreting ``GB`` as ``GiB`` (and vice versa) is clearly wrong and should be avoided. And the thing that for example LVM does where ``g`` means ``GiB`` and ``G`` means ``GB`` is really stupid and confusing.

Blivet's code for handling sizes is now factored out into [a standalone C library with GObject introspection] (https://github.com/rhinstaller/libbytesize) and so it would be really great if we could agree on some general policy and implement it there. Of course, the code should be shared by as many projects as possible, but I understand it's not an option for many.

**Comment by stefwalter** [comment]
@dperpeet I believe you're the one who was going to keep tabs on this discussion.

@vpodzime It's been the goal to be consistent with other UIs using the Patternfly toolkit. I think this is something that needs to be discussed with the Patternfly folks: http://patternfly.org/ -> #patternfly -> https://github.com/patternfly/patternfly

**Comment by vpodzime** [comment]
@stefwalter I know, thanks! I already tried to reach the PatternFly people via email and start a discussion with them. Haven't get too far yet, but I'll keep you guys informed if that changes.

**Comment by stefwalter** [comment]
@mvollmer Has opened a related pull request here #3944

**Comment by andreasn** [comment]
Since this is fixed in #3944, I'm closing this.
